add expose value parsers in OpamParser

gasche opened the pull request:
Currently, writing a parser from strings to values requires accessing
OpamBaseParser directly, which is a bit unpleasant -- it feels like
one is going too low-level. The proposed PR gives the same interface
to value parsers than to main/opamfile parsers.

I think another option would be to say that it is fine for users to
rely on OpamBaseParser directly (in particular, the unified interface
expects filename arguments that do not necessarily make much sense
when parsing values; maybe it is over-engineered), but then this
should be properly documented, for example by having
a OpamBaseParser.mli file versioned in the repository?

gasche committed:
add expose value parsers in OpamParser

The proposed PR gives the same interface to value parsers than to
main/opamfile parsers, and renames `value_of_string` into
`value_from_string` for consistency.

The new API takes filename arguments for the `value` parsers just like
file parsers. The optional argument `?loc` of the previous
`value_of_string` function is lost (it was not used in opam-ed, and
I don't know of any other users), but the raw `value` parser entry
point gives the same flexibility. A documentation example was added to
illustrate those.

AltGr commented:
Ah, thanks; somehow I only discovered your PR now, sorry about that !

gasche commented:
No problem. Am I correct that this is now subsumed by eb3d3431d7a9fcfbdfd982655e8bcd0bb39f4a96, and that this PR can be closed?

AltGr commented:
Well, your version was more complete :/
Anyway, the only uses (that I know of) are in [opam-ed](https://github.com/AltGr/opam-ed/blob/e1156e9b65f284c51fa850fde0bf1a3d10c2ec4f/src/opamEdMain.ml#L60). I'll close if you don't have time to resolve the conflicts.

gasche commented:
I'm happy to solve the conflicts, do you agree with the API proposed in the PR? You used `value_of_string`, while my PR used `value_from_string`, so I can move everything to `value_of_*` -- I think "value from channel" makes a bit more sense that "value of channel" because we are not converting the channel into a value, and this is the naming used in the Lexing and Scanf modules of the standard library, but I'm fine with `value_of`.

AltGr commented:
Yes; porting `opam-ed` won't be any trouble! Thanks :)

gasche commented:
I rebased the PR keeping the initial API proposed in the PR (do not hesitate to ask for something else). The `?loc` argument is gone, as I didn't find a nice way to include it. I think it was not used in `opam-ed` anyway. I added a documentation comment to show how to set the initial lexing position using the raw entry point directly.

AltGr commented:
Thanks a lot!
